Lwm2m queue mode DTLS resume

jheiskan81 opened the pull request:
- Enabled DTLS session cache for support session resume.
- Fixed LwM2M queue mode for close connection and reconnect automatically.
- Re-connect will do Registration update before it send queued data.
- Coap Confirmation pending class allocation fix.
- Registration update failure trig full Registration. This will handle session resume failure.

Commit:
net: lwm2m: DTLS session cache enable and queue mode update

Enabled DTLS session cache for support session resume.
Fixed LwM2M queue mode for close connection and reconnect automatically.
Re-connect will do Registration update before it send queued data.
Session resume is helping a case when NAT change address and cause less
network traffic.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
net: lwm2m: LwM2M RD client registration update failure fix

Client jump to full registration state if registration update fail.
Update keep already opened DTLS session.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
net: coap: Coap pending allocation fix

When sending 2 or more confirmable message before first one is
writed to socket all messages use same coap pending structure.
Now coap_pending_init() set data pointer which lock allocation
by each call.

Using data pointer for detecting free is more stable than timeout.
Timeout is initialized only before first socket send. Queued packet
may be triggered later than other and may cause that same block is
allocated multiple time.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos @SeppoTakalo @simensrostad @VeijoPesonen @jarlamsa Please review this.

rlubos reviewed (commented):
Session cache is still on a PR (https://github.com/zephyrproject-rtos/zephyr/pull/43710), feel free to review that to hasten the merge, the build will be failing before it's in.

Please address the compliance issue (Kconfig description), othherwise LGTM, already reviewed internally.

simensrostad reviewed (dismissed):
Looks good to me. Tested and confirmed working.

SeppoTakalo commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1147,7 +1147,13 @@ struct coap_pending *coap_pending_next_unused(
+		 * Timeout is initialized only before first socket send. Queued packet
+		 * may be triggered later than other and may cause that same block is allocated
+		 * multiple time.
+		 */
```
Remove the comment.
It is confusing to explain how things have been in the past.

It is obvious here, that if data is NULL, the structure is "free"

jheiskan81 commented:
@SeppoTakalo @rlubos @simensrostad I removed comment from Coap fix and added that to commit message. I also rebased  PR with upstream after PR #43710 is merged. Please review.
